pkg/restore/restore.go

```diff
@@ -1271,13 +1271,101 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
 					ctx.log.Infof("ServiceAccount %s successfully updated", kube.NamespaceAndName(obj))
 				}
 			default:
-				e := errors.Errorf("could not restore, %s %q already exists. Warning: the in-cluster version is different than the backed-up version.",
-					obj.GetKind(), obj.GetName())
-				warnings.Add(namespace, e)
+				// check for the presence of existingResourcePolicy
+				if len(ctx.restore.Spec.ExistingResourcePolicy) > 0 {
+					resourcePolicy := ctx.restore.Spec.ExistingResourcePolicy
+					ctx.log.Infof("restore API has resource policy defined %s , executing restore workflow accordingly for changed resource %s %s", resourcePolicy, fromCluster.GroupVersionKind().Kind, kube.NamespaceAndName(fromCluster))
+
+					// existingResourcePolicy is set as none, add warning
+					if resourcePolicy == velerov1api.PolicyTypeNone {
+						e := errors.Errorf("could not restore, %s %q already exists. Warning: the in-cluster version is different than the backed-up version.",
+							obj.GetKind(), obj.GetName())
+						warnings.Add(namespace, e)
+						return warnings, errs
+						// existingResourcePolicy is set as update, attempt patch on the resource and add warning if it fails
+					} else if resourcePolicy == velerov1api.PolicyTypeUpdate {
+						ctx.log.Infof("attempting patch on %s %q", fromCluster.GetKind(), fromCluster.GetName())
+						// remove restore labels so that we apply the latest backup/restore names on the object via patch
+						fromClusterWithLabels := fromCluster.DeepCopy() // saving the in-cluster object so that we can create label patch if overall patch fails
+						removeRestoreLabels(fromCluster)
+						patchBytes, err := generatePatch(fromCluster, obj)
+						if err != nil {
+							ctx.log.Infof("error generating patch for %s %s: %v", obj.GroupVersionKind().Kind, kube.NamespaceAndName(obj), err)
+							errs.Add(namespace, err)
+							return warnings, errs
+						}
+
+						if patchBytes == nil {
+							// In-cluster and desired state are the same, so move on to
+							// the next items
+							return warnings, errs
+						}
+
+						// try patching the in-cluster resource (resource diff plus latest backup/restore labels)
+						_, err = resourceClient.Patch(name, patchBytes)
+						if err != nil {
+							ctx.log.Infof("patch attempt failed for %s %s: %v", fromCluster.GroupVersionKind(), kube.NamespaceAndName(fromCluster), err)
+							warnings.Add(namespace, err)
+							// try just patching the labels
```

Comment:
I think we should also add this (patching just the labels if the patch attempt failed) to the ServiceAccount section above *if* the policy is update. Otherwise any user using prior restore labels to delete resources might delete SAs that didn't merge properly.

Reply:
@sseago added the patch labels bit if policy is update for SA too.